dhall/src/operations/typecheck.rs

```diff
@@ -503,7 +503,56 @@ pub fn typecheck_operation(
 
             selection_val
         }
-        Completion(..) | With(..) => {
+        With(record, labels, expr) => {
+            use crate::syntax::Label;
+            use std::iter::{once, FromIterator};
+
+            let record_entries = |nk: &NirKind| {
+                match nk {
+                    NirKind::RecordType(kts) => Ok(kts.clone()),
+                    _ => mk_span_err(span.clone(), "WithMustBeRecord"), // TODO better error
+                }
+            };
+
+            let mut current_nk: Option<NirKind> =
+                Some(record.ty().kind().clone());
+            let mut visited: Vec<(&Label, HashMap<Label, Nir>)> = Vec::new();
+            let mut to_create = Vec::new();
+
+            for label in labels {
+                match current_nk {
+                    None => to_create.push(label),
+                    Some(nk) => {
+                        let kts = record_entries(&nk)?;
+
+                        current_nk =
+                            kts.get(label).map(|nir| nir.kind().clone());
+                        visited.push((label, kts));
+                    }
+                }
+            }
+
+            // Create Nir for record type bottom up
+            let mut nir = expr.ty().as_nir().clone();
+
+            while let Some(label) = to_create.pop() {
+                let rec = RecordType(FromIterator::from_iter(once((
+                    label.clone(),
+                    nir,
+                ))));
+                nir = Nir::from_kind(rec);
+            }
+
+            // Update visited records
+            while let Some((label, mut kts)) = visited.pop() {
+                kts.insert(label.clone(), nir);
+                let rec = RecordType(kts);
+                nir = Nir::from_kind(rec);
+            }
+
+            nir.to_type(Const::Type)
```

Comment:
`with` can be used with records that contain types or kinds, so here `nir` may have a different type than `Type`. For example, the following is valid, and its type should be `{ x: Type }` which has type `Kind`.
```
{ x = Bool } with x = Natural
```
Here you want to take the `max` of the types of the entries of the last `kts`. But to compute that you need the type of the entries of the second-to-last `kts`, etc. So what you want is just `max` all the types you see, essentially.

Reply:
Oops actually `max(record.ty().ty().as_const(), expr.ty().ty().as_const())` should be enough

Reply:
Except I guess for
```
{ x = Bool } with x = 0
```
this decreases the kind of the record. Feel free to ignore that for now, but could you add those tests?

Reply:
Hah, I forgot I had some magic:
```suggestion
            Type::new_infer_universe(env, nir)
```
This should do everything I wanted. No need to keep track of things yourself.

Reply:
Actually, I tested several things and I think `nir.to_type(Const::Type)` works just fine. The record itself will always be a value even if the fields of the record contain types, kinds, or any other sorts.

It's definitely a value level record because of the `=` to bind fields to values. I'm not sure `with` would be possible/make sense on a type level record say `{ a : Bool } with ?`

Reply:
There's two different things here: one is record literals vs record types. Indeed `with` only applied to record literals. However a record literal may contain not only values like `True` and `0` but also values like `Bool` and `Type` (because in dhall everything is a value). Note the important distinction between `{ x = Bool }` and `{ x: Bool }`: one of them can be used with `with` and field access etc, but not the other.
So `{ a : Bool } with a = Natural` is indeed invalid. But `{ a = Bool } with a = Natural` is allowed and normalizes to `{ a = Natural }`.

Reply:
To be more concrete, if `record` is `RecordLit({ x = Bool })`, then `record.ty()` is `RecordType({ x : Type })` and `record.ty().ty()` is `Const(Kind)`. Here at the end you get `nir = RecordType(something)`, so the type to assign to `nir` should depend on `something`

Reply:
My understanding is that applying `with` always results in a `RecordLit(fields)`, so its type is always going to be the lowest level of `RecordType`. Or maybe there is something else I am missing. You are talking about doing `ty()` twice, but I don't think I do this while type checking.

Do you have an example that you think should break with the current implementation?

I have tried `{ x = Bool } with x = 0` and even`{ x = Bool } with x = Type` and they both seem to work just fine.

Reply:
When you return `nir.to_type(Const::Type)` what you're doing is say that the result of type inference is the type in `nir`, and you assert that its type is `Type`. This is not true if `nir` is `RecordType({ x : Type })` for example. For that one you should return `nir.to_type(Const::Kind)`. The reason we need to keep track of that is that several constructs only works for things of low kind, notably `toMap`.
Here is an example that passes with your code but shouldn't:
```dhall
toMap ({=} with x = Bool)
```

Reply:
Great thanks for the example! There are clearly some extra complexities around records that I need to understand better but I think I understand now that there are effectively multiple types of records at the value level. I do remember reading about that around the introduction of mixed type/value records but I completely forgot the details. Thank you for your patient explanations! :smile:

Reply:
No problem! It's definitely a tricky subject, not helped by the confusion of vocabulary between the type of a thing, the `Type` value, things that are types but are not `Type`, etc x)